### PR TITLE
API: Deprecate skip_footer in read_csv

### PR DESCRIPTION
**Pull request by gfyoung**
Title is self-explanatory.

Closes gh-13349 and partially undoes this <a href="https://github.com/pydata/pandas/commit/9ceea2fd46e79f37269a32de8c140cddf90eda13">commit</a> back in `v0.9.0`.  With such a massive API now, having duplicate arguments makes managing it way less practical.

### CODE REVIEW

**Comment by gfyoung** [comment]
In light of removing duplicate arguments and stripping down the API a bit, here are a couple of others I would like to propose (perhaps not in this PR but for discussion nonetheless):

1) Deprecate `skipfooter` in `read_excel` (it's allowed as alternative to `skip_footer`, which is surfaced in the function signature)?

2) Should we choose between `sep` and `delimiter` in `read_csv`?

**Comment by jreback** [comment]
the problem is that we also have ``skiprows``. I would actually deprecate ``skip_footer`` instead.

**Comment by gfyoung** [comment]
Hmmm...that's a flip-flop AFAICT from what you said earlier <a href="https://github.com/pydata/pandas/issues/13349#issuecomment-223450898">here</a>.  However, deprecating `skip_footer` seems a little strange since we use `skip_footer` all over the place in the code (internally especially).

**Comment by jreback** [comment]
just change it internally to the correct kw. I think they diverged at some point. It looks (just a quick skim), that ``skipfooter`` is more consistent (as we have ``skiprows``, ``usecols``) etc. excel should be fixed to be the same as well.

**Comment by gfyoung** [comment]
@jreback : let's actually do the count of **multi-word argument names** (excluding the `skip(_)footer` ones) here using the documentation as of `0.18.1` <a href="http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html">here</a>:

Arguments that use `_`:
`filepath_or_buffer`
`index_col`
`mangle_dupe_cols`
`true_values`
`false_values`
`na_values`
`keep_default_na`
`na_filter`
`skip_blank_lines`
`error_bad_lines`
`warn_bad_lines`
`delim_whitespace`
`as_recarray`
`compact_ints`
`use_unsigned`
`low_memory`
`buffer_lines`
`memory_map`
`float_precision`

Arguments that don't use `_`:
`usecols`
`skipinitialspace`
`skiprows`
`chunksize`
`lineterminator`
`doublequote`

Even excluding ones that have been recently deprecated (`compact_ints`, `buffer_lines`, `use_unsigned`, `as_recarray`), arguments that use `_` vastly outnumber those that don't.

**Comment by jreback** [comment]
@gfyoung I am trying to minimize back-compat pain.

``usecols`` and ``skiprows`` are prob some of the most used options. I agree we should just deprecate and move to ``_`` separated, but that would be in 0.19.0 if we go whole hog.

**Comment by gfyoung** [comment]
@jreback : I have no issues waiting until `0.19.0` and going full hog.  This duplicate argument is not breaking, and then it will give us the time to properly deprecate and rename arguments with `_`, which I foresee could apply to these functions:

`read_csv`
`read_excel`
`read_html`

**Comment by gfyoung** [comment]
@jreback : I have no issues waiting until `0.19.0` and going full hog.  This duplicate argument is not breaking, and it will then give us the time to properly deprecate and rename arguments with `_`, which I foresee could apply to these functions:

`read_csv`
`read_excel`
`read_html`

**Comment by gfyoung** [comment]
Rebased onto `master` and Travis is passing.  I see that @jorisvandenbossche has added the "Needs Discussion" tag, so comments are welcome!

File: doc/source/io.rst [hunk 1]
```diff
@@ -282,16 +285,16 @@ float_precision : string, default None
   Specifies which converter the C engine should use for floating-point values.
   The options are ``None`` for the ordinary converter, ``high`` for the
   high-precision converter, and ``round_trip`` for the round-trip converter.
-lineterminator : str (length 1), default ``None``
```
Comment (jreback) [inline doc/source/io.rst]: so anything you are actually deprecating (which is pretty much everything that is changed), needs to have a either a separate entry for a doc-string (like you did for ``skip_footer``), or prob better, show the doc-string for ``skip_footer``  then at the end put the DEPRECATION OF ``skipfooter`` or somesuch. 

Futher need a BIG BIG warning in v0.19.0 (use ``::warning``).

File: doc/source/whatsnew/v0.19.0.txt [hunk 2]
```diff
@@ -439,6 +439,17 @@ Deprecations
 - ``buffer_lines`` has been deprecated in ``pd.read_csv()`` and will be removed in a future version (:issue:`13360`)
 - ``as_recarray`` has been deprecated in ``pd.read_csv()`` and will be removed in a future version (:issue:`13373`)
 - top-level ``pd.ordered_merge()`` has been renamed to ``pd.merge_ordered()`` and the original name will be removed in a future version (:issue:`13358`)
+- The ``chunksize``, ``dayfirst``, ``doublequote``, ``escapechar``, ``lineterminator``, ``quotechar``, ``skipinitialspace``,
```
Comment (jreback) [inline doc/source/whatsnew/v0.19.0.txt]: make this a separate section. add to the highlites (single line with reference to this section).

File: doc/source/whatsnew/v0.19.0.txt [hunk 3]
```diff
@@ -439,6 +439,17 @@ Deprecations
...
+``skipfooter``, ```skiprows``, and ``usecols`` arguments in ``pd.read_csv()`` have been deprecated in favor of their
+underscore counterparts ``chunk_size``, ``day_first``, ``double_quote``, ``escape_char``, ``line_terminator``, ``quote_char``,
+``skip_initial_space``, ``skip_footer``, ```skip_rows``, and ``use_cols`` arguments respectively (:issue:`13386`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.19.0.txt]: make this a mini-table for easier viewing.

File: pandas/io/parsers.py [hunk 4]
```diff
@@ -116,13 +116,16 @@
...
-skipinitialspace : boolean, default False
+skip_initial_space : boolean, default False
     Skip spaces after delimiter.
-skiprows : list-like or integer, default None
+skip_rows : list-like or integer, default None
     Line numbers to skip (0-indexed) or number of lines to skip (int)
     at the start of the file
-skipfooter : int, default 0
+skip_footer : int, default 0
     Number of lines at bottom of file to skip (Unsupported with engine='c')
```
Comment (jreback) [inline pandas/io/parsers.py]: whatever you do in the docs needs to be the same here

**Comment by jreback** [comment]
this looks good to me. need to have a nice way of communicating the deprecations in the doc-string. typically we add an additional entry for the new one and make the original kwarg have a DEPRECATED in bold. Here we need to do something different as so many args.

maybe just list the new ones, but have a sentence that says the orginal arg is deprecated. 

AND have a section before the arguments start that says, hey all of these are deprecated.

finally, need to create an issue (maybe just a single one), to have ``read_excel``, ``read_html``, ``read_hdf`` made consistent with this (I don't think any others, but should check). e.g. ``chunksize`` for sure is one.

**Comment by jreback** [comment]
also have issue #4988 which is kind of what I am suggesting for above (e.g. its about ``read_excel``); so maybe repurpose that one to be the non-csv-parser but same kw-in-need-of-deprecation issue.

File: doc/source/io.rst (hunk 1 above)
Comment (gfyoung) [inline doc/source/io.rst]: Can you illustrate with an example?  I don't quite understand what you're saying here.

File: pandas/io/parsers.py (hunk 4 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: Can you clarify?  The reason why `skipfooter` is slightly different from the rest is because there were already duplicate args for these two, whereas the others don't.

**Comment by gfyoung** [comment]
@jreback : Agreed about `read_excel` - that has a few arguments that could be changed, though best saved for another PR.  I was planning on looking those afterwards after working out the kinks with this one here.

File: doc/source/whatsnew/v0.19.0.txt (hunk 3 above)
Comment (gfyoung) [inline doc/source/whatsnew/v0.19.0.txt]: How do I make a table in these files?

File: doc/source/whatsnew/v0.19.0.txt (hunk 2 above)
Comment (gfyoung) [inline doc/source/whatsnew/v0.19.0.txt]: Made a separate section, but need help on the table part.

File: doc/source/whatsnew/v0.19.0.txt [hunk 5]
```diff
@@ -436,6 +438,15 @@ Deprecations
 - top-level ``pd.ordered_merge()`` has been renamed to ``pd.merge_ordered()`` and the original name will be removed in a future version (:issue:`13358`)
 - ``Timestamp.offset`` property (and named arg in the constructor), has been deprecated in favor of ``freq`` (:issue:`12160`)
 
+.. _whatsnew_0190.deprecations.read_csv_arg_rename:
+
+All multi-word arguments to ``pd.read_csv()`` will now have their individual words separated by an underscore, and their
+ underscore-less versions have not be deprecated (:issue:`13386`). They are:
+
+- The ``chunksize``, ``dayfirst``, ``doublequote``, ``escapechar``, ``lineterminator``, ``quotechar``, ``skipinitialspace``,
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.19.0.txt]: @jreback : GitHub unfortunately hid away the other comment about the table, so I'm "refreshing it."  Could you explain how to create tables in these files?

Comment (jreback) [inline doc/source/whatsnew/v0.19.0.txt]: look in 0.18.0 whatsnew - there is a table about msgpack

Comment (gfyoung) [inline doc/source/whatsnew/v0.19.0.txt]: For future reference, the example you are referring to is <a href="https://github.com/pydata/pandas/blob/master/doc/source/whatsnew/v0.18.0.txt#L614">here</a>.  Ah, got it.  Done.

**Comment by jorisvandenbossche** [comment]
The reason I added "Needs discussion" is because I am not sure if I find this a good idea. `read_csv` is one of the most used pandas functions, and causing thousands/millions (?) users to have to change their code just for 'aesthetic'* reasons, not sure this is worth the trouble.

\* I know this is not *only* aesthetic, because having a consistent API (eg always underscores) also makes it easier to learn, to remember, etc .. And that is very important, but sometimes 'historical' reasons are good reasons.

**Comment by jorisvandenbossche** [comment]
On the specific keywords:

- `dayfirst` is also used in `to_datetime`, `Timestamp`, and comes from `dateutil.parser.parse` , so I would leave this is as.
- `chunksize` is also used a lot in other places (`read_hdf`, `read_sql`, `to_sql`, ..), so I would also be more hesitant to change this

**Comment by jreback** [comment]
an alternative is simply to accept aliases, e.g. ``skiprows`` & ``skip_rows``. This only adds a very tiny amount of code (in getting kwargs). and 'solves' the problem pretty easily.

**Comment by jorisvandenbossche** [comment]
Yes, you could see it as a "deprecation by documentation": documenting the recommended way, but not actually deprecating (and maybe a small note the docstring that certain list of keywords are also accepted for historical reasons). 
This options of course also has some overhead and disadvantages (possible confusion when users see code using the 'old' args)

**Comment by gfyoung** [comment]
@jorisvandenbossche :

`chunksize` is one that @jreback and I discussed earlier in the conversation as converting to `chunk_size` (note the list that I made <a href="https://github.com/pydata/pandas/pull/13386#issuecomment-231613825">here</a>).

As for `dayfirst`, I do realize that there are some inconsistencies in deprecating it, especially since `datetime` uses that parameter.  However, a consistent API is useful (i.e. all underscores), so similar to what @jreback is suggesting, we could just document only the recommended parameters but still accept the old parameters instead with `DeprecationWarning` and the recommended parameters (as I do now).  However, I can add a `TODO` or comment saying that we should NOT remove support for the old parameters for some time.

How does that sound?

**Comment by gfyoung** [comment]
@jorisvandenbossche :

`chunksize` is one that @jreback and I discussed earlier in the conversation as converting to `chunk_size` (note the list that I made <a href="https://github.com/pydata/pandas/pull/13386#issuecomment-231613825">here</a>).

As for `dayfirst`, I do realize that there are some inconsistencies in deprecating it, especially since `datetime` uses that parameter.  However, a consistent API is useful (i.e. all underscores), so similar to what @jreback is suggesting, we could just document only the recommended parameters but still accept the old parameters with `DeprecationWarning` and message about the recommended parameters (as I do now).  However, I can add a `TODO` or comment saying that we should NOT remove support for the old parameters for some time.

How does that sound?

**Comment by gfyoung** [comment]
@jorisvandenbossche :

`chunksize` is one that @jreback and I discussed earlier in the conversation as converting to `chunk_size` (note the list that I made <a href="https://github.com/pydata/pandas/pull/13386#issuecomment-231613825">here</a>).

As for `dayfirst`, I do realize that there are some inconsistencies in deprecating it, especially since `datetime` uses that parameter.  However, a consistent API is useful (i.e. all underscores), so similar to what @jreback is suggesting, we could just document only the recommended parameters but still accept the old parameters with `DeprecationWarning` and message about the recommended parameters (as I do now).  However, I can add a `TODO` or comment saying that we should NOT remove support for the old parameters for some time.

How does that sound?

File: doc/source/cookbook.rst [hunk 6]
```diff
@@ -998,12 +998,12 @@ Skip row between header and data
     01.01.1990 05:00;21;11;12;13
     """
 
-Option 1: pass rows explicitly to skiprows
+Option 1: pass rows explicitly to skip_rows
 """"""""""""""""""""""""""""""""""""""""""
```
Comment (jreback) [inline doc/source/cookbook.rst]: if you change the len, you need to change the underlining as well.

**Comment by jreback** [comment]
@gfyoung nice table!

but I agree with @jorisvandenbossche I don't think we should deprecate these keywords. It really not a big deal to simply accept both in the constructors (and document it in the doc-strings).

There aren't that many. What I mean is change the ``pd.read_csv`` to accept ``line_terminator``, but also allow ``lineterminator`` (you will have to do a check that *both* are not defined), but this can be automated. I DO think that we should change all internal code / docs to use the underscore separated ones (as you have done). So most of your docs can stay. 

I just don't think we should actually deprecated them (but instead deprecate in the future) after this has been out there for a bit.

kind of like a PendingDeprecationWarning.

**Comment by gfyoung** [comment]
@jreback : Thanks!  It's amazing what you can do when all you have to do is copy. :smile: 

Hmm...I guess it's because I've been through this on the `numpy` side when I deprecated `random_integers` but the plan wasn't to remove it for the foreseeable future.  I guess it depends on how you interpret deprecation.

File: doc/source/cookbook.rst (hunk 6 above)
Comment (gfyoung) [inline doc/source/cookbook.rst]: Good catch.  Will fix once Travis is happy.

**Comment by jreback** [comment]
well, let's not get started on what ``numpy`` does or doesn't do :)

**Comment by gfyoung** [comment]
Well technically, it's what I did, not `numpy`. ;)

**Comment by jreback** [comment]
we don't use ``PendingDeprecationWarning``. I would just have an issue that schedules us to do it. though not really against ``PendingDeprecationWarning``

**Comment by gfyoung** [comment]
Hmmm...given that it's such a massive change, I wouldn't mind warning users in advance of the deprecation.

**Comment by jreback** [comment]
yeah I guess that is appropriate, even python itself uses them: https://github.com/numpy/numpydoc/issues/35

@jorisvandenbossche ?

**Comment by gfyoung** [comment]
@jreback : <a href="https://github.com/pydata/pandas/pull/13386#issuecomment-232235928">hypocrite</a>!

File: doc/source/io.rst [hunk 7]
```diff
@@ -168,21 +168,24 @@ true_values : list, default ``None``
...
-skiprows : list-like or integer, default ``None``
+skip_rows : list-like or integer, default ``None``
   Line numbers to skip (0-indexed) or number of lines to skip (int) at the start
   of the file.
-skipfooter : int, default ``0``
+skip_footer : int, default ``0``
   Number of lines at bottom of file to skip (unsupported with engine='c').
+skipfooter : int, default ``None``
+  DEPRECATED: this argument will be removed in a future version. Please
+  use the ``skip_footer`` parameter instead.
```
Comment (jorisvandenbossche) [inline doc/source/io.rst]: Why did you add this one? (given you did not add it for the others, which is good I think)

Comment (gfyoung) [inline doc/source/io.rst]: `skipfooter` and `skip_footer` were already in the signature before this PR, hence the necessity to document both.

File: doc/source/io.rst (hunk 1 above)
Comment (jorisvandenbossche) [inline doc/source/io.rst]: I would add a note (``.. note:: ``) somewhere on this page (maybe add the end of the enumeration of all the options), explaining that for some keywords aliases without underscores are accepted for historical reasons, but that it is recommended to use the version with underscores.

File: pandas/io/parsers.py [hunk 8]
```diff
@@ -116,13 +116,16 @@
...
+skipfooter : int, default None
+    DEPRECATED: this argument will be removed in a future version. Please
+    use the `skip_footer` parameter instead.
```
Comment (jorisvandenbossche) [inline pandas/io/parsers.py]: Same comment here: why add it for skipfooter?

In the docstring, I would add a similar note about other keyword arguments without underscores being accepted (can be the same as in io.rst) in the 'Notes' section.

Comment (gfyoung) [inline pandas/io/parsers.py]: See my response <a href="https://github.com/pydata/pandas/pull/13386#discussion_r70590017">above</a>.

File: doc/source/io.rst (hunk 7 above)
Comment (jorisvandenbossche) [inline doc/source/io.rst]: Personally I don't think that is a reason to do differentely for this one (in the end, for all the others, the non-underscore versions were *also* in the signature)

Comment (gfyoung) [inline doc/source/io.rst]: I suppose if we're going to keep accepting both versions for ALL of them, that is a fair point.  Will remove.

**Comment by jorisvandenbossche** [comment]
> but I agree with @jorisvandenbossche I don't think we should deprecate these keywords. It really not a big deal to simply accept both in the constructors (and document it in the doc-strings).

Yes that is how I see it as well.

> kind of like a PendingDeprecationWarning.

So then just let's use one :-) But the questions is maybe: are we ever going to actually deprecate them? If not, we maybe don't need to use any warning (but also in that case you could see the PendingDeprecationWarning as a kind of NotRecommendedKeywordPleaseUseOtherWarning :-), so maybe no problem with having that). 
In any case, I would not use a FutureWarning

> As for ``dayfirst``, I do realize that there are some inconsistencies in deprecating it, especially since datetime uses that parameter. However, a consistent API is useful.

-1 on changing the `dayfirst` kwarg. An inconsistent naming scheme within a function is not good, but having inconsistent names for the same thing between functions is even worse IMO. So either we should rename it everywhere, or either keep it here as `dayfirst`. 
But personally my preference is just to keep it as `dayfirst` (given that it corresponds with a dateutil kwarg). Having one keyword that deviates from the naming scheme is not that a big deal I think.

> > ``chunksize`` is also used a lot in other places (``read_hdf``, ``read_sql``, ``to_sql``, ..), so I would also be more hesitant to change this

> chunksize is one that @jreback and I discussed earlier in the conversation as converting to chunk_size.

Maybe it is my Dutch mother tongue that I find `chunksize` good as one word :-) (in Dutch, such a compound word would be one word instead of two separate words)

Since this is a rather big change (regarding user impact), I think it is good to have some more eyes on it. cc @TomAugspurger @sinhrks @shoyer @wesm @chris-b1 If you have any objections/opinions on the general idea, time to raise them.

**Comment by jreback** [comment]
I would say leave chunksize and dayfirst alone as joris indicated

File: doc/source/io.rst (hunk 1 above)
Comment (gfyoung) [inline doc/source/io.rst]: @jorisvandenbossche : I'm not sure how viable that is given that the enumeration is split into different sections.

**Comment by shoyer** [comment]
I would lean against this change -- I don't think deprecating or even adding aliases with underscores for these keywords is worth it. Yes, consistency is nice, but these are keywords that are mostly either (a) copied straight from the API docs or an example or (b) auto-completed. Omitting underscores between lower case words is also common enough in PEP8 compliant code that it doesn't hurt my eyes in the way that camelCase would, for example. I would save changes like this for next time somebody writes a DataFrame library ;). There's just very little upside to the change at this point.

**Comment by chris-b1** [comment]
I'd also be inclined to not do this.

Consistency is great, but to me, big breaking (or warning triggering) changes for consistency are only worth it to the extent they actually improve the user experience.  E.g. `resample`, the new `sort_...` api, or  a closer example, when the inconsistency between the `rows` and `index` keyword was resolved in `pivot_table` / `pivot` - #5505

This doesn't seem to meet that hurdle.  Like @shoyer said, these argument would be typically tab-completed or looked up, or even if not, the inconsistency doesn't cause any confusion.

**Comment by gfyoung** [comment]
To put things in perspective, this whole discussion was generated because I wanted to remove the duplicate 'skipfooter/skip_footer' argument in the signature.  It really should only be one of the two IMO.

The consistency part came up because we weren't sure which way to go i.e do we take the underscore version or the non-underscore?

I perfectly understand if we don't want to make such a change as drastic as it has become, though some input on the initial goal would be good then.

**Comment by chris-b1** [comment]
I don't have a strong opinion on that, but given that all the docs / SO answers / etc in the wild use `skipfooter`, I'd think deprecating `skip_footer` makes more sense?

**Comment by gfyoung** [comment]
In light of the push back, I will rollback my changes and just deprecate `skip_footer` as @chris-b1 indicated.  If `skipfooter` is being used in the wild, let's stick with that one then.

**Comment by gfyoung** [comment]
@jreback , @jorisvandenbossche : Rolled back my changes successfully to just deprecating `skip_footer` for the reason that @chris-b1 outlined above.  Ready to merge if there are no other concerns.

**Comment by jreback** [comment]
lgtm. @jorisvandenbossche

**Comment by gfyoung** [comment]
@jorisvandenbossche : any updates on this?

**Comment by jorisvandenbossche** [comment]
What do we do other `read_` functions for the skipfooter case?

**Comment by gfyoung** [comment]
Same as here pre-PR, I will handle all other cases similarly if this one is merged.

**Comment by gfyoung** [comment]
@jorisvandenbossche : Any updates on this?

**Comment by gfyoung** [comment]
@jorisvandenbossche : Any updates on this?

**Comment by gfyoung** [comment]
@jorisvandenbossche : Any updates on this?

File: doc/source/io.rst [hunk 9]
```diff
@@ -1351,7 +1351,7 @@ back to python if C-unsupported options are specified. Currently, C-unsupported
 options include:
 
 - ``sep`` other than a single character (e.g. regex separators)
-- ``skip_footer``
+- ``skipfooter``
```
Comment (jreback) [inline doc/source/io.rst]: did the doc-string get changed (to add DEPRECATED)?

Comment (gfyoung) [inline doc/source/io.rst]: Hmmm...not sure where those changes went.   Added them back.

**Comment by gfyoung** [comment]
@jreback : added the documentation back about the deprecation, and Travis is still happy after rebase.  Ready to merge if there are no other concerns.

**Comment by jreback** [comment]
thanks @gfyoung
